### PR TITLE
Support for swarm and JSON input

### PR DESCRIPTION
**Pull request by apoorva11029**
In this commit:
- Testing support for Swarm with Vagrant and Baremetal
- Support for ACI and non-ACI mode on Baremetal.
- Input from a single JSON file for the entire framework.

### CODE REVIEW

**Comment by gaurav-dalvi** [comment]
LGTM. Good design overall !

**Comment by gaurav-dalvi** [comment]
@jainvipin , @shaleman @jojimt : Could you please take a look at this one ?

This is systemtest framework so that now we can run system tests on baremetal as well.

**Comment by gaurav-dalvi** [comment]
@apoorva11029 : Could you please squash all commits in to one commit and the submit PR again ?

**Comment by abhinandanpb** [comment]
@apoorva11029 have u tested these changes with l3 test and k8 test ?

**Comment by apoorva11029** [comment]
@abhinandanpb thanks for reminding. Done testing L3 and k8s for Vagrant.

**Comment by abhinandanpb** [comment]
@apoorva11029 please watch for this jenkins run for failures for l3 and k8s
http://contiv.ngrok.io/view/Netplugin/job/Netplugin_OnDemand_Full_Sanity/152/console

**Comment by apoorva11029** [comment]
@abhinandanpb the JSON input needs to be modified for k8s/L3

**Comment by abhinandanpb** [comment]
@apoorva11029 the design should be such that - when respective tests are run then corresponding config files should be generated.

File: test/systemtests/aci_test.go [hunk 1]
```diff
@@ -72,403 +73,582 @@ func (s *systemtestSuite) TestACIMode(c *C) {
...
 	c.Assert(s.cli.EndpointGroupDelete("default", "epgb"), IsNil)
-	c.Assert(s.cli.NetworkDelete("default", "aciNet"), IsNil)
+	c.Assert(s.cli.NetworkDelete("default", s.globInfo.Network), IsNil)
 }
 
-/*
+/* TesACIPingGateway checks ping success from containers running in a EPG to the default gateway */
 func (s *systemtestSuite) TestACIPingGateway(c *C) {
-	if s.fwdMode == "routing" {
+	if s.fwdMode == "routing" || s.basicInfo.Scheduler == "k8" {
```
Comment (abhinandanpb) [inline test/systemtests/aci_test.go]: the reason k8 check was added was because the code was not in sync with the new framework. this can be removed now.

File: test/systemtests/aci_test.go [hunk 2]
```diff
@@ -72,403 +73,582 @@ func (s *systemtestSuite) TestACIMode(c *C) {
...
+}
 
+/* TestACIProfile does the following:
+1). TestACIMode and TestACIPingGateway for containers in two EPGs.
+2). Checks policies and rules learned on the APIC controller.
+3). Checks policies behavior before and after deleting app-profiles.
+*/
 func (s *systemtestSuite) TestACIProfile(c *C) {
-	if s.fwdMode == "routing" {
+	if s.fwdMode == "routing" || s.basicInfo.Scheduler == "k8" {
```
Comment (abhinandanpb) [inline test/systemtests/aci_test.go]: same as above

File: test/systemtests/aci_test.go [hunk 3]
```diff
@@ -72,403 +73,582 @@ func (s *systemtestSuite) TestACIMode(c *C) {
...
-	for i := 0; i < 2; i++ {
-		log.Infof(">>ITERATION #%d<<", i)
+	containersA := []*container{}
+	containersB := []*container{}
+
+	containersA2 := []*container{}
+	containersB2 := []*container{}
+	for i := 0; i < s.basicInfo.Iterations; i++ {
+
 		c.Assert(s.cli.NetworkPost(&client.Network{
```
Comment (abhinandanpb) [inline test/systemtests/aci_test.go]: i am assuming the config file will have option to include multiple networks. This may need to be handled here looping over the number of networks listed in the config file. This can be part of separate PR. Please open a bug once the PR is merged

File: test/systemtests/aci_test.go [hunk 4]
```diff
@@ -72,403 +73,582 @@ func (s *systemtestSuite) TestACIMode(c *C) {
...
+			NetworkName: s.globInfo.Network,
+			Subnet:      s.globInfo.Subnet,
+			Gateway:     s.globInfo.Gateway,
+			Encap:       s.globInfo.Encap,
 		}), IsNil)
 
+		err := s.nodes[0].checkSchedulerNetworkCreated(s.globInfo.Network, false)
+		c.Assert(err, IsNil)
+
 		c.Assert(s.cli.EndpointGroupPost(&client.EndpointGroup{
```
Comment (abhinandanpb) [inline test/systemtests/aci_test.go]: similarly for epgs. if we have multiple epgs.

File: test/systemtests/aci_test.go [hunk 5]
```diff
@@ -72,403 +73,582 @@ func (s *systemtestSuite) TestACIMode(c *C) {
...
+		containersB = nil
+		c.Assert(s.cli.EndpointGroupDelete(s.globInfo.Tenant, "epga"), IsNil)
+		c.Assert(s.cli.EndpointGroupDelete(s.globInfo.Tenant, "epgb"), IsNil)
+		c.Assert(s.cli.RuleDelete(s.globInfo.Tenant, "policyAB", "2"), IsNil)
+		c.Assert(s.cli.RuleDelete(s.globInfo.Tenant, "policyAB", "3"), IsNil)
+
+		c.Assert(s.cli.PolicyDelete(s.globInfo.Tenant, "policyAB"), IsNil)
+		c.Assert(s.cli.NetworkDelete(s.globInfo.Tenant, s.globInfo.Network), IsNil)
+	}
+}
```
Comment (abhinandanpb) [inline test/systemtests/aci_test.go]: Ideally all the ACI tests should conform to the existing tests. 
If we have taken the config file approach. we need to mention the physical networks information. Which should be used to iterate and derive the number of epgs,networks policies.  This can be part of a seperate PR. file an issue for this after PR is merged

File: test/systemtests/cfg.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
+[
```
Comment (abhinandanpb) [inline test/systemtests/cfg.json]: This adds a lot of flexibility to our tests which is great. We will need to be able to either auto generate/auto fill or create static config files upfront. So depending the kind of testSuite running corresponding file can be generated/used. This can be driven by ENV variables like CONTIV_L3, CONTIV_K8,CONTIV_MESOS,CONTIV_ACI etc.

File: test/systemtests/swarm_test.go [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+		dnsStr = "--dns=" + spec.dnsServer
+	}
+
+	if len(spec.labels) > 0 {
+		l := "--label="
+		for _, label := range spec.labels {
+			labelstr += l + label + " "
+		}
+	}
+	logrus.Infof("Starting a container running %q on %s", spec.commandName, w.node.Name())
```
Comment (abhinandanpb) [inline test/systemtests/swarm_test.go]: similar to k8. It might be good to maintain the node where the container is scheduled on. some use cases for k8s.  like for ping test you can show container on which node is trying to ping. This log is printed in the currents tests. But this will be very useful info for debugging

File: test/systemtests/swarm_test.go [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+	logrus.Infof("Expecting ping failure from %v to %s", c, ipaddr)
+	if err := w.checkPing(c, ipaddr); err == nil {
+		return fmt.Errorf("Ping succeeded when expected to fail from %v to %s", c, ipaddr)
+	}
+
+	return nil
+}
+
+func (w *swarm) checkPing(c *container, ipaddr string) error {
+	logrus.Infof("Checking ping from %v to %s", c, ipaddr)
```
Comment (abhinandanpb) [inline test/systemtests/swarm_test.go]: Here from %v will be the printing the info i mentioned in the above comment.

File: test/systemtests/swarm_test.go [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+	for _, ep := range epList {
+		if !strings.Contains(str, ep) {
+			return str, errors.New(ep + " not found on " + w.node.Name())
+		}
+	}
+
+	return "", nil
+}
+
+func (w *swarm) reloadNode(n *node) error {
```
Comment (abhinandanpb) [inline test/systemtests/swarm_test.go]: This is currently not being used because the test case skipped. This will need an abstraction for a node. Since this is very specific to vagrant. Please file an issue once PR is merged.

File: test/systemtests/util_test.go [hunk 10]
```diff
@@ -902,3 +906,302 @@ func (s *systemtestSuite) verifyEPs(containers []*container) error {
...
+			Vxlans:           "1-10000",
+		}), IsNil)
+		time.Sleep(40 * time.Second)
+	}
+}
+
+func (s *systemtestSuite) SetUpTestVagrant(c *C) {
+	for _, node := range s.nodes {
+		node.exec.cleanupContainers()
+		//node.exec.cleanupDockerNetwork()
```
Comment (abhinandanpb) [inline test/systemtests/util_test.go]: is this needed ?

**Comment by abhinandanpb** [comment]
The changes look great overall. This brings in lot of flexibility to system tests. Thanks for taking care of it. It will be great if the config file scenarios are finalized. After that please open issues to track post merge.

File: scripts/python/createcfg.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+data['mgmtInterface'] = args.mgmtinterface
+data['vlan'] = args.vlan
+data['vxlan'] = args.vxlan
+data['subnet'] = args.subnet
+data['gateway'] = args.gateway
+data['network'] = args.network
+data['tenant'] = args.tenant
+data['encap'] = args.encap
+
+filepath = os.environ['GOPATH'] + '/src/github.com/contiv/netplugin/test/systemtests/cfg.json'
```
Comment (gaurav-dalvi) [inline scripts/python/createcfg.py]: It would good to display some message saying that ,

"generating config file required for systest-framework"

File: scripts/python/createcfg.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python
+
+import os 
+import json
+import argparse 
+
```
Comment (gaurav-dalvi) [inline scripts/python/createcfg.py]: Good write comment saying:

This script aims to generate the config json file which has all these fields.

**Comment by gaurav-dalvi** [comment]
this will be merged via this PR : https://github.com/contiv/netplugin/pull/555

File: test/systemtests/aci_test.go (hunk 1 above)
Comment (gaurav-dalvi) [inline test/systemtests/aci_test.go]: Removed in new PR !

File: test/systemtests/aci_test.go (hunk 3 above)
Comment (gaurav-dalvi) [inline test/systemtests/aci_test.go]: opened up issue for this : https://github.com/contiv/netplugin/issues/567

File: test/systemtests/aci_test.go (hunk 4 above)
Comment (gaurav-dalvi) [inline test/systemtests/aci_test.go]: opened up issue for this : https://github.com/contiv/netplugin/issues/567

File: test/systemtests/aci_test.go (hunk 5 above)
Comment (gaurav-dalvi) [inline test/systemtests/aci_test.go]: opened up issue for this : https://github.com/contiv/netplugin/issues/567

File: test/systemtests/cfg.json (hunk 6 above)
Comment (gaurav-dalvi) [inline test/systemtests/cfg.json]: createcfg.py takes care of creating json file on the fly. Changed at places in makefile as well.

File: test/systemtests/swarm_test.go (hunk 7 above)
Comment (gaurav-dalvi) [inline test/systemtests/swarm_test.go]: Opened up issue https://github.com/contiv/netplugin/issues/567 for this.

File: test/systemtests/swarm_test.go (hunk 9 above)
Comment (gaurav-dalvi) [inline test/systemtests/swarm_test.go]: Opened up issue for this : https://github.com/contiv/netplugin/issues/567

File: test/systemtests/util_test.go (hunk 10 above)
Comment (gaurav-dalvi) [inline test/systemtests/util_test.go]: Will refactor code more in next PR

**Comment by gaurav-dalvi** [comment]
As this was closed via https://github.com/contiv/netplugin/pull/555. Closing this PR now.
